python/kvikio/kvikio/_lib/buffer.pyx

```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+# distutils: language = c++
+# cython: language_level=3
+
+
+from kvikio._lib.arr cimport Array
+
+
+cdef extern from "<kvikio/buffer.hpp>" namespace "kvikio" nogil:
+    void cpp_memory_register "kvikio::memory_register"(const void* devPtr) except +
+    void cpp_memory_deregister "kvikio::memory_deregister"(const void* devPtr) except +
+
+
+def memory_register(buf) -> None:
+    if not isinstance(buf, Array):
+        buf = Array(buf)
+    cdef Array arr = buf
+    cpp_memory_register(<void*>arr.ptr)
+
+
+def memory_deregister(buf) -> None:
+    if not isinstance(buf, Array):
+        buf = Array(buf)
+    cdef Array arr = buf
+    cpp_memory_deregister(<void*>arr.ptr)
```

Comment:
Can you help me understand the model here? Does Array(buf) create a view or a copy? Does deregistration of memory require ownership, like a free? (I don’t really know what registration means.) I’m trying to understand how it’s legal to take an input buffer that isn’t an array, create an array, and deregister that new array’s pointer.

Reply:
Sounds good.